FWCore/Catalog/src/InputFileCatalog.cc

```diff
@@ -14,46 +17,54 @@ namespace edm {
-        overrideFileLocator_(),
-        fallbackFileLocator_(),
-        overrideFallbackFileLocator_() {
-    init(override, "", useLFNasPFNifLFNnotFound);
-  }
-
-  InputFileCatalog::InputFileCatalog(std::vector<std::string> const& fileNames,
-                                     std::string const& override,
-                                     std::string const& overrideFallback,
-                                     bool useLFNasPFNifLFNnotFound)
-      : logicalFileNames_(fileNames),
-        fileNames_(fileNames),
-        fallbackFileNames_(fileNames.size()),
-        fileCatalogItems_(),
-        fileLocator_(),
-        overrideFileLocator_(),
-        fallbackFileLocator_(),
-        overrideFallbackFileLocator_() {
-    init(override, overrideFallback, useLFNasPFNifLFNnotFound);
+      : logicalFileNames_(fileNames), fileNames_(fileNames), fileCatalogItems_(), overrideFileLocator_() {
+    init(override, useLFNasPFNifLFNnotFound);
   }
 
   InputFileCatalog::~InputFileCatalog() {}
 
-  void InputFileCatalog::init(std::string const& inputOverride,
-                              std::string const& inputOverrideFallback,
-                              bool useLFNasPFNifLFNnotFound) {
-    fileCatalogItems_.reserve(fileNames_.size());
+  std::vector<std::string> InputFileCatalog::fileNames(unsigned iCatalog) const {
+    std::vector<std::string> tmp;
+    tmp.reserve(fileCatalogItems_.size());
+    for (auto const& item : fileCatalogItems_) {
+      tmp.push_back(item.fileName(iCatalog));
+    }
+    return tmp;
+  }
+
+  void InputFileCatalog::init(std::string const& inputOverride, bool useLFNasPFNifLFNnotFound) {
     typedef std::vector<std::string>::iterator iter;
-    for (iter it = fileNames_.begin(),
-              lt = logicalFileNames_.begin(),
-              itEnd = fileNames_.end(),
-              ft = fallbackFileNames_.begin();
-         it != itEnd;
-         ++it, ++lt, ++ft) {
+
+    if (!overrideFileLocator_ && !inputOverride.empty()) {
+      overrideFileLocator_ =
+          std::make_unique<FileLocator>(inputOverride);  // propagate_const<T> has no reset() function
+    }
+
+    Service<SiteLocalConfig> localconfservice;
+    if (!localconfservice.isAvailable())
+      throw cms::Exception("TrivialFileCatalog", "edm::SiteLocalConfigService is not available");
+
+    std::vector<std::string> const& tmp_dataCatalogs = localconfservice->dataCatalogs();
+    if (!fileLocators_.empty())
+      fileLocators_.clear();
+
+    for (auto it = tmp_dataCatalogs.begin(); it != tmp_dataCatalogs.end(); ++it) {
+      //require the first file locator to success so obvious mistakes in data catalogs, typos for example, can be catched early
+      if (it == tmp_dataCatalogs.begin()) {
+        fileLocators_.push_back(std::make_unique<FileLocator>(*it));
+      } else {
+        try {
+          fileLocators_.push_back(std::make_unique<FileLocator>(*it));
+        } catch (cms::Exception const& e) {
+          continue;
+        }
```

Comment:
How about the following (to avoid repeated checks inside the loop)?
```suggestion
    //require the first file locator to success so obvious mistakes in data catalogs, typos for example, can be catched early
    fileLocators_.push_back(std::make_unique<FileLocator>(tmp_dataCatalogs.front()));
    for (auto it = tmp_dataCatalogs.begin()+1; it != tmp_dataCatalogs.end(); ++it) {
      try {
        fileLocators_.push_back(std::make_unique<FileLocator>(*it));
      } catch (cms::Exception const& e) {
        continue;
      }
```
(if `tmp_dataCatalogs` can be empty at this point, then a protection would be needed)

This is my only remaining comment.

Reply:
Hi @makortel , yes this is better. Should I wait for all others to approve or comment before making another "push" with modification suggested by you and others? I am just afraid that we go to many iteration if I make individual "push". At the previous iteration, dqm, reconstruction and generator gave approval. Those codes are not touched this time.

Reply:
Just do the push, I'm pretty sure the others will sign after that (or I would have expected them to comment already if they would have any).